File: solutions/0024_swap-nodes-in-pairs.c
// Problem: https://leetcode.com/problems/swap-nodes-in-pairs
// Runtime: 3 ms

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* swapPairs(struct ListNode* head) {
    int index = 0;
    struct ListNode* ret = head;
    struct ListNode* last = NULL;
    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    struct ListNode* next = NULL;
    while (curr != NULL) {
        // printf("%d\n", curr->val);
        next = curr->next;
        if (index % 2 == 1) {
            prev->next = curr->next;
            curr->next = prev;
            if (last != NULL) {
                last->next = curr;
            }
            last = prev;
            if (index == 1) {
                ret = curr;
            }
        }
        prev = curr;
        curr = next;
        index++;
    }
    return ret;
}

File: solutions/0024_swap-nodes-in-pairs.py
# Problem: https://leetcode.com/problems/swap-nodes-in-pairs
# Runtime: 40 ms

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        last2 = []
        curr = head
        ret = head
        
        index = 0
        while curr != None:
            nextNode = curr.next
            if index % 2 == 1:
                last2[-1].next = nextNode
                curr.next = last2[-1]
                if len(last2) > 1:
                    last2[-2].next = curr
                if index == 1:
                    ret = curr
                last2.append(last2[-1])
            else:
                last2.append(curr)
            if len(last2) > 2:
                last2.pop(0)
            curr = nextNode
            index += 1

        return ret